File: src/B_2512/B_2512.java
package B_2512;

import java.util.*;
import java.io.*;

public class B_2512{
    public static void main(String[] args)throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int N = Integer.parseInt(st.nextToken());
        int[] arr = new int[N];

        st = new StringTokenizer(br.readLine()," ");
        for(int i=0; i<N; ++i){
            arr[i] = Integer.parseInt(st.nextToken());
        }
        long M = Integer.parseInt(br.readLine());
        Arrays.sort(arr);
        long start = 1;
        long end = arr[N-1];
        long res = 0;
        long mid =0;
        while(start <= end){
            mid = (start+end)/2;
            long sum = 0;
            for(int i=0; i<N; ++i){
                sum += Math.min(arr[i],mid);

            }
            if(sum <= M){
                start = mid + 1;
                res = mid;
            }else{
                end = mid - 1;

            }

        }
        System.out.print(res);
    }
}

File: src/B_2512/B_2512.kt
package B_2512

import java.io.*
import java.util.*

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine(), " ")
    val N = st.nextToken().toInt()
    val arr = IntArray(N)
    st = StringTokenizer(br.readLine(), " ")
    for (i in 0 until N) {
        arr[i] = st.nextToken().toInt()
    }
    val M = br.readLine().toInt().toLong()
    Arrays.sort(arr)
    var start: Long = 1
    var end = arr[N - 1].toLong()
    var res: Long = 0
    var mid: Long = 0
    while (start <= end) {
        mid = (start + end) / 2
        var sum: Long = 0
        for (i in 0 until N) {
            sum += Math.min(arr[i], mid.toInt())
        }
        if (sum <= M) {
            start = mid + 1
            res = mid
        } else {
            end = mid - 1
        }
    }
    print(res)
}